### DISCUSSION TITLE
Can't make SerialBridge to work

### DISCUSSION THREAD

**Discussion by 19cool53** [arendst/Tasmota#13829] [Q&A] [answered]
By reading older posts I was given the opinion that the serial bridge should be able to "accept" a json encoded string.
This would give the advantage of sending parameter data that looks alike the RFreceived message on a Sonoff bridge.
S I tried with tasmota.bin v9.4.0.
Sending String :  {"SerID":"TM182", "MyCmd":"on"}         // no <lf>  or <cr> was included
Tasmota received:
{"SSerialReceived" : {"SerID":"TM182", "MyCmd":"on"} }            // as expected
rule1 on SSerialReceived#SerID=TM182 do POWER1 toggle endon    // did not work

below: xxxx = POWER1 toggle

Sending String :  TM182              
Received result:  {"SSerialReceived" : TM182 }   // and rule //   on SSerialReceived=TM182 do xxx endon     // did not work

After updating to version 10.0, behaviour is different: (added \ before the " )

  Sending String :  {"SerID":"TM182", "MyCmd":"on"}                        // no <lf>  or <cr> was included
 {"SSerialReceived" : {"SerID":"\"TM182\"","\"MyCmd\"":"\"on\""} }    // that's a bit weird and does not work

  Sending String :  { SerID:TM182, MyCmd:on }                                 // no, not json at all
 {"SSerialReceived":{ SerID:TM182, MyCmd:on } }                       }    // no, complete nonsens

Sending String :  TM182              
Received result:  {"SSerialReceived" : TM182 }   // and rule //   on SSerialReceived=TM182 do POWERxxx endon     // did not work

Sending String :  "TM182"              
Received result:  {"SSerialReceived" : \"TM182\" }   // and rule //   on SSerialReceived=TM182 do xxx endon     // did not work

In fact, the only thing I try to do is - analog to a sonoff bridge -
- sending a device-id
- receive the device-id over serial
- trigger a rule event if the device-id matches.

So what's am I doing wrong? Where is the missing link? Can anybody give me a hint?

Regards, cool

**Comment by barbudor** [comment, marked as answer]
Using your initial example, it works fine:
```
16:34:15.765 CMD: rule1 on SSerialReceived#sid==TM154 do var1 sid=%value% endon
16:34:15.770 RUL: Stored uncompressed, would compress from 55 to 43 (-22%)
16:34:15.776 MQT: stat/nodemcu/RULE = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":55,"Free":456,"Rules":"on SSerialReceived#sid==TM154 do var1 sid=%value% endon"}}
16:34:34.852 MQT: tele/nodemcu/SSERIALRECEIVED = {"SSerialReceived":{"Data":"set","sid":"TM154","MyCmd":"power1 toggle"}}
16:34:34.866 RUL: SSERIALRECEIVED#SID==TM154 performs "var1 sid=TM154"
16:34:34.872 MQT: stat/nodemcu/VAR = {"Var1":"sid=TM154"}
```

Now testing your other exemples

**Reply by barbudor** [reply]
1st example above ok:
```
16:38:13.368 CMD: rule1 on SSerialReceived#SerID=TM182 do POWER1 toggle endon
16:38:13.372 RUL: Stored uncompressed, would compress from 53 to 43 (-19%)
16:38:13.377 MQT: stat/nodemcu/RULE = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":53,"Free":458,"Rules":"on SSerialReceived#SerID=TM182 do POWER1 toggle endon"}}
16:38:21.548 MQT: tele/nodemcu/SSERIALRECEIVED = {"SSerialReceived":{"SerID":"TM182", "MyCmd":"on"}}
16:38:21.562 RUL: SSERIALRECEIVED#SERID=TM182 performs "POWER1 toggle"
16:38:21.566 MQT: stat/nodemcu/POWER = {"POWER":"ON"}
```

Just by any change, did you enabled the rule with `Rule1 1` ?
Please provide output of command `Rule0`  (no space between `Rule` and `0`)
And the copy of console logs when you send one of your JSON string

**Reply by sfromis** [reply]
The reason why `on SSerialReceived=TM182 do POWERxxx endon` did not work should be that a JSON trigger for "historical reasons" need to be suffixed by `#data` when the JSON only has a single value, like:
`on SSerialReceived#data=TM182 do POWERxxx endon`

And `{"SSerialReceived" : TM182 }` is just not valid JSON.

**Reply by barbudor** [reply]
This one use a wrong syntax:
> Sending String : TM182
> Received result: {"SSerialReceived" : TM182 } // and rule // on SSerialReceived=TM182 do xxx endon // did not work

Correct syntax is `on SSerialReceived#data=TM182`. When the JSON response has only 1 field, you must use `#data` 
```
16:43:33.214 CMD: rule1 on SSerialReceived#data=TM182 do POWER1 toggle endon
16:43:33.219 RUL: Stored uncompressed, would compress from 52 to 41 (-21%)
16:43:33.223 MQT: stat/nodemcu/RULE = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":52,"Free":459,"Rules":"on SSerialReceived#data=TM182 do POWER1 toggle endon"}}
16:44:07.834 MQT: tele/nodemcu/SSERIALRECEIVED = {"SSerialReceived":"TM182"}
16:44:07.850 RUL: SSERIALRECEIVED#DATA=TM182 performs "POWER1 toggle"
16:44:07.857 MQT: stat/nodemcu/POWER = {"POWER":"OFF"}
``` 

I'm testing with latest 10.0.0.3

**Reply by barbudor** [reply]
I also noticed that 1 element that may the JSON parsing fails is the string received from serial not starting immedialy with a `{`
Tasmota checks for that `{` as 1st character to decide that it is a JSON string and parse it as a JSON

**Comment by 19cool53** [comment]
Great! Thx a lot.

Here is my result:


Works:   (TM182 without "")
--------------------------------------
send : TM182
recv:  "SSerialReceived":"TM182"
rule:   rule1 on SSerialReceived#Data=TM182 do xxxx endon

           rule1 1   ;)


works also:
---------------
send : {"sid":"TM182" }
recv:  "SSerialReceived":{"sid":"TM182" }
rule:   rule1 on SSerialReceived#sid=TM182 do xxxx endon


but this does not:
-----------------------
send: {"sid":"TM182", "svar":"xxx" }
recv:   "SSerialReceived":{"sid":"TM182","svar":"xxx" }
rule:    rule1 on SSerialReceived#sid=TM182 do xxxx endon


however, this works
---------------------------
send : { "svar":"xxx", "sid":"TM182" }
recv:  "SSerialReceived":{ "svar":"xxx",  "sid":"TM182" }
rule:   rule1 on SSerialReceived#sid=TM182 do xxxx endon


Looks to be the main reason (plus the #Data issue) why I was
suffering. The matching JSON item should be the last provided.

So far this is working. Thanks.

My remaining questions:
----------------------------------
1) How can I receive data from a serial device? A trigger needs
     to match ( #Data=xxx or #TM182=xxx). This will not work
    if the value changes  (like sensor data)

2) The var1 sid=%value% matches the same as the trigger.
     If I know the trigger, I don'tt need the value.

What I need is a trigger on #key and var1 = %key value%
and that does not seem to be there (at least not
documented)

Regards
cool


Am 27.11.21 um 16:40 schrieb Barbudor:

**Reply by barbudor** [reply]
Your 3rd case works like a charm:

```
21:58:38.088 CMD: rule1 on SSerialReceived#sid=TM182 do var1 %value% endon
21:58:38.093 RUL: Stored uncompressed, would compress from 50 to 39 (-22%)
21:58:38.098 MQT: stat/nodemcu/RULE =
{"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":50,"Free":461,"Rules":"on
SSerialReceived#sid=TM182 do var1 %value% endon"}}
21:58:59.583 MQT: tele/nodemcu/SSERIALRECEIVED =
{"SSerialReceived":{"sid":"TM182", "svar":"xxx" }}
21:58:59.598 RUL: SSERIALRECEIVED#SID=TM182 performs "var1 TM182"
21:58:59.605 MQT: stat/nodemcu/VAR = {"Var1":"TM182"}
```

As you don't post the exact console logs it is difficult to say why, but I
would assume that `{` was not the 1st character of the serial string.

> 1) How can I receive data from a serial device? A trigger needs
>  to match ( #Data=xxx or #TM182=xxx). This will not work
>  if the value changes  (like sensor data)
> 2) The var1 sid=%value% matches the same as the trigger.
>   If I know the trigger, I don'tt need the value.

You use `=TM182` only if you want the trigger to occur when this value
occurs.
If you want to trigger for any JSON that has a `sid` key and get the value
of that key, just use
`ON SSerialReceived#sid DO somewith_with %value% ENDON`
Like :

```
22:03:39.341 CMD: rule1 on SSerialReceived#sid do var1 %value% endon
22:03:39.346 RUL: Stored uncompressed, would compress from 44 to 31 (-30%)
22:03:39.351 MQT: stat/nodemcu/RULE =
{"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":44,"Free":467,"Rules":"on
SSerialReceived#sid do var1 %value% endon"}}
22:04:20.652 MQT: tele/nodemcu/SSERIALRECEIVED =
{"SSerialReceived":{"sid":"THX1138", "svar":"xxx"}}
22:04:20.664 RUL: SSERIALRECEIVED#SID performs "var1 THX1138"
22:04:20.669 MQT: stat/nodemcu/VAR = {"Var1":"THX1138"}

```

**Comment by 19cool53** [comment]
Tanks for your prompt answer.
Tnaks to your hints, I found this post and a working solution:

https://github.com/arendst/Tasmota/issues/8232#

|Rule1
     on SerialReceived#pressure do backlog var1 %value%; knxtx_val1 
%value% endon
     on SerialReceived#temp do backlog var2 %value%; knxtx_val2 %value% 
endon
     on SerialReceived#status do backlog var3 %value%; knxtx_val3 
%value% endon
     on SerialReceived#error do backlog var4 %value%; knxtx_val4 %value% 
endon|

That allows grabbing even multiple values. That's perfect and covers
my 1st question and makes 2) obsolete.


What I have to regret is my the lack of understanding, that makes it
very difficult to figure out how things (should) work. The docs are
helpfull, but many times I do not understand what is meant and
what dependencies exist. Simpe examples that can be rebuild are
great, maybe a Wiki can collect experience of those many thousend
Tasmota users, suffering the same?
Any way, thanks to your hints we're getting closer.

You mentioned the console logs. These would not show what has been
sent, but here is the difference between json or simple text inputs:

sent:    TM182:25.3
08:54:37.235 MQT: tele/Sensors/tasmota_B92950/RESULT = 
{"SSerialReceived":"TM182:25.3"}

sent:    "TM182":"25.3"
08:56:06.997 MQT: tele/Sensors/tasmota_B92950/RESULT = 
{"SSerialReceived":"\"TM182\":\"25.3\""}

sent:    {"TM182":"25.3"}
08:59:04.000 MQT: tele/Sensors/tasmota_B92950/RESULT = 
{"SSerialReceived":{"TM182":"25.3"}}

sent:   _{"TM182":"25.3"}
09:00:15.901 MQT: tele/Sensors/tasmota_B92950/RESULT = 
{"SSerialReceived":" {\"TM182\":\"25.3\"}"}



So I may use something like this (as a working example):

sent:   {"SerID":"TM182","T_182":"25.3", "L_182":"50%" }
09:26:07.302 MQT: tele/Sensors/tasmota_B92950/RESULT = 
{"SSerialReceived":{"SerID":"TM182","T_182":"25.3", "L_182":"50" }}

rule1 1
rule1  on SSerialReceived#serid    do  var1 %value% endon
           on SSerialReceived#t_182   do  publish 
cmnd/Sensors/%var1%/Outside_Temperature %value% endon
           on SSerialReceived#l_182   do  publish 
cmnd/Sensors/%var1%/Outside_Light  %value% endon

09:51:44.114 MQT: tele/Sensors/tasmota_B92950/RESULT = 
{"SSerialReceived":{"SerID":"TM182","T_182":"25.3", "L_182":"50" }}
09:51:44.124 RUL: SSERIALRECEIVED#SERID performs "var1 TM182"
09:51:44.130 MQT: stat/Sensors/tasmota_B92950/RESULT = {"Var1":"TM182"}
09:51:44.141 RUL: SSERIALRECEIVED#T_182 performs "publish 
cmnd/Sensors/TM182/Outside_Temperature 25.3"
09:51:44.144 MQT: cmnd/Sensors/TM182/Outside_Temperature = 25.3
09:51:44.155 RUL: SSERIALRECEIVED#L_182 performs "publish 
cmnd/Sensors/TM182/Outside_Light  50"
09:51:44.160 MQT: cmnd/Sensors/TM182/Outside_Light =  50


Looks like my neural computer got a few synapses linked  ;)

Much oblied, thanks and:
Can I put this example somewhere?


Regards
Cornelis



Am 27.11.21 um 22:53 schrieb Barbudor:

**Reply by Jason2866** [reply]
Doc is OpenSource to. You are invited to add examples ;-)
Just click the pencil in upper right corner.

**Reply by barbudor** [reply]
What I meant with the logs is that you can see exactly what Tasmota as received
Back to your example with the string `{"sid":"TM182", "svar":"xxx" }`
Depending if you send that properly with `{` as the 1st character you have such log:
```
12:15:20.275 MQT: tele/nodemcu/SSERIALRECEIVED = {"SSerialReceived":{"sid":"TM182", "svar":"xxx" }}
```
Now if you mis-send adding a space in front of the ` {` you would get such log:
```
12:15:49.879 MQT: tele/nodemcu/SSERIALRECEIVED = {"SSerialReceived":" {\"sid\":\"TM182\", \"svar\":\"xxx\" }"}
```
Notice the difference.
You may consider that adding 1 space at the beginning is not a problem but it definitively changes the way Tasmota understand the string : 
- in the 1st case, Tasmota parse it as a JSON and all fields are reachable with `#`
- in the 2nd case, it just remains a string where you can't access the fields

So when you wrote:
> but this does not:
> -----------------------
> send: {"sid":"TM182", "svar":"xxx" }
>recv:   "SSerialReceived":{"sid":"TM182","svar":"xxx" }
>rule:    rule1 on SSerialReceived#sid=TM182 do xxxx endon

without providing the exact log, I can't tell why it doesn't work in your case while I can prove it works on my side.

Providing exact information and logs is important to get proper answers in support.


Regarding docs, as Jason wrote, you can propose addition and changes
Nevertheless I just added this section : https://tasmota.github.io/docs/Rules/#processing-json-received-from-softwareserialbridge
Feel free to add more if you feel it will help.

**Comment by 19cool53** [comment]
Thanks for the hint.
While searching where to put it, I saw that Barbadour allready putted 
what we
figured out (see: json over serial), but it's good to know.
I've learned quite a lot, thanks to all.

Am 28.11.21 um 12:17 schrieb Jason2866:
